tests/e2e/collectives/run_requirements.txt

```diff
@@ -0,0 +1,2 @@
+mpi4py==3.1.4
```

Comment:
On my Linux (Debian) machine:

```shell
python -m pip install -r tests/run_requirements.txt

...
      _configtest.c:2:10: fatal error: 'mpi.h' file not found
      #include <mpi.h>
               ^~~~~~~
      1 error generated.
      failure.
      removing: _configtest.c _configtest.o
      error: Cannot compile MPI programs. Check your configuration!!!
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for mpi4py
Failed to build mpi4py
ERROR: Could not build wheels for mpi4py, which is required to install pyproject.toml-based projects
```

Is that a hard requirement for the Python test? Does the runtime C code need similar dependencies?

Some ideas:

* We could use https://cmake.org/cmake/help/latest/module/FindMPI.html to enable/disable the test 
* https://openxla.github.io/iree/building-from-source/getting-started/#dependencies could mention `python -m pip install -r tests/run_requirements.txt` for running tests in addition to `python -m pip install -r runtime/bindings/python/iree/runtime/build_requirements.txt` for building the python bindings
  * I'd prefer for the run requirements to install cleanly without additional system package installs though
* We can jump through some hoops in our Docker images as needed (e.g. installing Open MPI or whatever else https://pypi.org/project/mpi4py/ wants), but the developer experience for the core project should work without a hard requirement to install specific requirements (e.g. CUDA, Vulkan SDK, etc.)
* Does pip support conditional requirements? If not, maybe we shouldn't put a direct include between `tests/run_requirements.txt` and `tests/e2e/collectives/run_requirements.txt` and just keep it optional. Not great either way :/

Reply:
Maybe we could replace `rank = MPI.COMM_WORLD.Get_rank()` with some call into IREE's Python API, and have that use some underlying C code that doesn't require an extra package install?

Reply:
I added a CMake flag to exclude collectives tests.

> * Does pip support conditional requirements? If not, maybe we shouldn't put a direct include between `tests/run_requirements.txt` and `tests/e2e/collectives/run_requirements.txt` and just keep it optional.

Pip requirements files seem to not have conditional control of what should be installed.
Conditional test run environment setup should be driven by the configuration of the build. It would probably need to generate the installation procedure.
Do you want me to include such facilities in this PR? Right now the pip requirements for the collectives tests are not included anywhere.

> Maybe we could replace rank = MPI.COMM_WORLD.Get_rank() with some call into IREE's Python API, and have that use some underlying C code that doesn't require an extra package install?

This approach will work during build time. In IREE the MPI library is dynamically loaded with `dlopen`.
During runtime you would still need `libmpi.so`. There is not pip package for that. At least OpenMPI does not have one.

It seems NCCL can be installed with pip
https://pypi.org/project/nvidia-nccl-cu11/
https://pypi.org/project/nvidia-nccl-cu12/